File: schedule-service/src/main/java/mrsisk/github/io/appointservice/security/GrantedAuthoritiesExtractor.java
package mrsisk.github.io.appointservice.security;

import org.springframework.core.convert.converter.Converter;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.oauth2.jwt.Jwt;

import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

class GrantedAuthoritiesExtractor
        implements Converter<Jwt, Collection<GrantedAuthority>> {

    public Collection<GrantedAuthority> convert(Jwt jwt) {
        Map<?, ?> auth = (Map<?, ?>)
                jwt.getClaims().getOrDefault("realm_access", Collections.emptyMap());

        var authorities = auth.get("roles") == null ? Collections.emptyList() : (List<?>) auth.get("roles") ;

        return authorities.stream()
                .map(Object::toString)
                .map(SimpleGrantedAuthority::new)
                .collect(Collectors.toList());
    }
}

File: accounts-service/src/main/kotlin/mrsisk/github/io/accountsmanager/security/GrantedAuthoritiesExtractor.kt
package mrsisk.github.io.accountsmanager.security

import org.springframework.core.convert.converter.Converter
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.oauth2.jwt.Jwt

internal class GrantedAuthoritiesExtractor : Converter<Jwt, Collection<GrantedAuthority>> {
    override fun convert(jwt: Jwt): Collection<GrantedAuthority> {
        val authorities = (jwt.claims
            .getOrDefault("realm_access", emptyMap<String, Any>()) as Map<*, *>)

        val roles = (authorities["roles"] ?: emptyList<String>()) as List<*>

        return roles
            .map { it.toString() }
            .map { SimpleGrantedAuthority(it) }
    }
}